src/client/core/aad/auth-provider.ts

```diff
@@ -63,45 +79,83 @@ export default class AuthProvider {
-            if (result) {
-                this._accounts[tenantId] = result.account;
-            }
             return result;
         }
     }
 
-    public logout(): void {
-        this._removeAccount();
+    public async logout(): Promise<void> {
+        this._logoutPromise = this._removeAccounts();
+        return this._logoutPromise;
+    }
+
+    private async _removeAccounts(): Promise<void> {
+        const cache = this._primaryClient?.getTokenCache();
+        if (!cache) {
+            return;
+        }
+        const accounts = await cache.getAllAccounts();
+        for (const account of accounts) {
+            await cache.removeAccount(account);
+        }
+        this._accounts = {};
+        this._clients = {};
+        this._primaryClient = undefined;
     }
 
     protected _getClient(tenantId: string): ClientApplication {
         if (tenantId in this._clients) {
             return this._clients[tenantId];
         }
-        return this._clients[tenantId] = new PublicClientApplication({
+        const client = new PublicClientApplication({
             auth: {
                 clientId: this.config.clientId,
                 authority:
                     `${this.app.properties.azureEnvironment.aadUrl}${tenantId}/`
+            },
+            cache: {
+                cachePlugin: this._cachePlugin
             }
         });
+        if (!this._primaryClient) {
+            this._primaryClient = client;
+        }
+        this._clients[tenantId] = client;
+        return client;
     }
 
-    private async _removeAccount(): Promise<void> {
-        for (const tenantId in this._clients) {
-            if (this._accounts[tenantId]) {
-                const cache = this._clients[tenantId].getTokenCache();
-                cache.removeAccount(this._accounts[tenantId]);
+    private async _getAccount(tenantId: string): Promise<AccountInfo> {
+        if (tenantId in this._accounts) {
+            return this._accounts[tenantId];
+        }
+        const cache = this._clients[tenantId].getTokenCache();
+        const accounts: AccountInfo[] = await cache.getAllAccounts();
+        let homeAccountId = null;
+        for (const account of accounts) {
+            if (account.tenantId === tenantId) {
+                return this._accounts[tenantId] = account;
+            } else if (!homeAccountId) {
+                homeAccountId = account.homeAccountId;
             }
-            delete this._accounts[tenantId];
         }
+
+        /* SPECIAL CASE: If the tenant is one of the tenant placeholders (e.g.,
+         * "common"), fallback to the home tenant account, since the tenant in
+         * the account dictionary are always resolved IDs.
+         */
+        if (tenantId in TenantPlaceholders) {
+            return this._accounts[tenantId] =
+                await cache.getAccountByHomeId(homeAccountId);
+        }
+
+        throw `Unable to find a valid AAD account for tenant ${tenantId}`;
```

Comment:
Just for consistency let's throw an error rather than a string (makes it easier to not have to do type checking when catching errors)